design/service-apis-implementation.md

```diff
@@ -0,0 +1,326 @@
+Contour will have an entry added to the config file for the Gateway it should watch.
+
+This will look like this:
+```yaml
+gateway:
+  name: gatewayname
+  namespace: gatewaynamespace
+```
+
+The current flag `--experimental-service-apis` will be removed.
+The boolean it represents will be deduced from the following rules:
+- the gateway is configured in the config file.
+- the referenced gateway exists.
+
+If the gateway is configured but does not exist, then this is a fatal error and Contour will exit.
+If the gateway is removed while Contour is operating, then an error will be logged, and all config associated with the Service APIs will be removed from Envoy.
+
+### Code changes
+Contour already has support for importing the Service APIs objects into its Kubernetes cache.
+However, for some types, we need to be able to keep some details - this design suggests making those details properties of the cache, as `IngressClass` is currently.
+
+For ingestion, the general pattern currently is that the EventHandler in `internal/contour/handler.go` handles all objects, and calls out to the KubernetesCache from `internal/dag/cache.go`.
+The current pattern across most of the objects is:
+- check if the object is in scope
+- add it to the cache
+- indicate with a return value if the add should result in a DAG (and consequently an Envoy config) rebuild.
+
+This pattern is applicable to all the Service APIs objects, with the exception of GatewayClass, which is not relevant for Contour.
+
+There are already internal fields in the cache to hold GatewayClass and some other service-apis objects.
+GatewayClass will be removed as part of this work, then we will use this scaffolding to build the actual functionality on top.
+
+#### Hostname matching
+
+In many of the Service APIs objects, Hostnames *may* be specified, and *may* be either "precise" (a domain name without the terminating dot of a network host), or "wildcard" (a domain name prefixed with a single wildcard label).
+Per the API spec, only the first DNS label of the hostname may be a wildcard, and the wildcard must match only a single label.
+
+Hostnames are considered to match if they exactly match, or if a precise hostname matches the suffix of a wildcard hostname (that is, if they match after the first DNS label is discarded.)
+
+This DNS matching is referred to in this proposal as a "domain match", as opposed to an "exact match", which is string equality only.
+
+### Gateway
+
+When ingesting Gateways, Contour will import the configured named Gateway into its cache, and watch it for spec changes. Spec changes will trigger a DAG run.
+
+#### Listener Ports
+
+Contour allows the configuration of arbitrary ports, **as long as those ports are exposed via the Envoy service**.
+The port that is configured for the listener is the port as it appears at the Service.
+That is, the Envoy Service `port` number is what the Listener must specify, regardless of if there is a `targetPort` set on the Service.
```

Comment:
Maybe a comment somewhere making it explicit that we will start watching the Envoy service would be good to add. Also it forces users (cluster operators?) to deploy their Envoys in a certain way. Practically speaking its unlikely someone had their Envoys deployed much differently than we have in the examples or the operator, but it is technically a new constraint I guess? Maybe not that serious

Reply:
I've updated this quite a bit to reflect the fact that we're not going to allow arbitrary ports for now. There's definitely a problem that will need to be solved with some experimentation during implementation here.